_posts/2019-12-17-zero-to-main-rust-1.md

```diff
@@ -0,0 +1,433 @@
+---
+title: "From zero to main(): Bare metal Rust"
+description: "Bringing an embedded Rust application from zero to main"
+author: james
+tags: [zero-to-main]
+---
+
+For the past thirty years or so, the choices of languages for embedded systems developers has been relatively slim. Languages like C++ and Ada have found a home in some niche areas, such as telecommunications and safety critical fields, while higher level languages like Lua, Python, and JavaScript have found a home for scripting and prototyping. But most developers working on bare metal systems have been using the same two languages as long as I can remember: Assembly and C.
+
+But not for no reason! Languages often make trade-offs to fit the needs of the developers working with them: an interpreter to allow for rapid iteration, a heap for ease of memory management, exceptions for simplifying control flow, etc. But these trade-offs come with a price: whether it is code size, RAM usage, low level control, power usage, latency, or determinism.
+
+<!-- excerpt start -->
+
+Since 2015, Rust has been redefining what it means to combine the best-in-class aspects of performance, correctness, and developer convenience into one language, without compromise.
+
+<!-- excerpt end -->
+
+As a compiled systems language (based on LLVM), it is also capable of reaching down to the lowest levels of embedded programming as well, without losing built-in features that feel more at home in higher level languages, like a package manager, helpful compile time diagnostics, correctness through powerful static analysis, or documentation tooling.
+
+In this series of posts, we'll show how to bootstrap a Rust environment from scratch, integrate with existing C codebases, write correct code that is enforced at compile time, and more!
+
+This post is meant as a complement to the original [Zero to Main] post on the Interrupt blog, and will elide some of the explanations of hardware level concepts. If you're new to embedded development, or haven't seen that post, go read it now!
+
+[Zero to Main]: https://interrupt.memfault.com/blog/zero-to-main-1
```

Comment:
```suggestion
[Zero to Main]: {{ '/tag/zero-to-main' |
relative_url  }}
```

Reply:
In this case, I want to link directly to the first post, rather than the series (as that one covers all the stuff I skip in this post). Would you prefer that I still link to the series? Or use some kind of template to link to that post?

Reply:
Got it. What you have is good.